executor: Introduce Executor "Trait Objects" by using Type Erasure

retrage opened the pull request:
<!-- Set a title that summarizes the PR changes. -->

## Type of PR
<!-- This project uses issue-driven development, so please take the appropriate action for each PR type. -->

- [x] Changes related to the roadmap (e.g., TODO.md) (type: A) -> Create an issue corresponding to the PR in advance, and refer to this PR in the issue.
- Changes that are not related to the roadmap
    - [ ] Change with multiple possible solutions to the issue (type: B-1) -> Create an issue corresponding to the PR in advance and refer to this PR in the issue.
    - [ ] Change with a single solution (type: B-2) -> There is no need to create an issue corresponding to the PR in advance. Please discuss it in this PR.

## Related Issue
<!-- If this PR is a PR type A/B-1, please provide a link to the corresponding issue. -->
<!-- If this PR is a PR type B-1, please write "N/A" -->
#30 

## Importance of PR
<!-- Please describe the importance of the PR in terms of the following aspects. -->

- Importance of the issue
    - [ ] Large (based on several days to weeks of discussion and verification, e.g., this issue is a blocking issue for other issues on the roadmap, etc.)
    - [x] Medium (based on a few hours to a day of discussion and verification, e.g., this issue is a blocking issue for another minor issue)
    - [ ] Small (apparent changes such as build error)
- Complexity of the solution (code, tests, etc.)
    - [ ] Large (requires several days to several weeks of review)
    - [x] Medium (requires several hours to a day of review)
    - [ ] Small (trivial changes, such as build error)

## PR Overview
<!-- Please provide a summary of this PR. -->
<!-- If this PR is a PR type A/B-1, this PR will be considered as an item in the checklist for the related issue. Please provide a link to the issue comment that contains the checklist. -->
<!-- If this PR is a PR type B-2, unnecessary to reference the issue. Please provide a summary. -->
For the background, see #30.
This PR proposes introducing "trait object" in Rust's term [1] to represent what methods must be implemented for an executor to support the algorithm. By this change, the algorithms can use executors without knowing the actual executor type since the classes abstract the executors. I implemented the introduced classes (`DynAFLCapable` and `DynLibFuzzerCapable`) using type erasure as suggested by @potetisensei. Note that this PR adds bare-type erasure classes only, so it does not affect the behaviors of the algorithm and CLI. I'm going to create another PR that introduces CLI changes for selecting executors at runtime to show the advantage of this change.

[1] https://doc.rust-lang.org/book/ch17-02-trait-objects.html

## Concerns (Optional)
<!-- If you have any concerns, please describe them clearly by filling in the relevant checklist items below. If there is anything else you would like to share with the reviewer, please include it. -->

- [x] Performance
- [x] Source Code Quality

I have several concerns about this change:

### 1. Type Erasure Implementation

I implemented the classes with type erasure on my own. This is usually discouraged because bugs in my own implementation of type erasure are difficult to debug and can lead to "shooting myself in the foot". However, I found that the implementation using boost/type_erasure did not meet our needs in this case, as the amount of code using boost was not much different from my own implementation, and error messages tended to be more difficult. For this reason, I decided to implement type erasure myself. There is room for discussion on this policy.

### 2. Performance

Because type erasure hides the actual type, runtime performance may suffer depending on compile-time optimizations. Although I have not yet measured and compared performance, I did not experience any performance degradation when running the unit test.

Please review the type erasure classes to ensure that no unnecessary copies are made. They take `std::shared_ptr<TExecutor>` as a constructor argument rather than executor itself. I believe that the executor is not created unnecessarily, but I am not sure if this implementation is correct.

### 3. Naming

These classes are placed under `include/fuzzuf/executor/traits`. I used the `traits` directory name because it is designed to be aware of Rust's trait objects, but it may cause misunderstanding because trait in C++ is a different concept from Rust's trait. If you have a better alternative directory name, please comment.

---

> The PR author should fill in the following checklist when submitting this PR.

#### Optional Entries
- [x] If this PR is a PR type A/B-1, there is a cross-link between this PR and the related issue.

#### Mandatory Entries
- [x] The PR title is a summary of the changes.
- [x] Completed each required field of the PR.

---
> The PR author should fill out the following checklist in the comments to confirm that this PR is ready to be merged

- [ ] CI is green or confirmed test run results.
- [ ] All change suggestions from reviewers have been resolved (fixed or foregone).

---
> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [ ] The reviewer assigned more reviewers if needed.
- [ ] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [ ] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [ ] The title of this PR summarizes the changes made by this PR properly.
- [ ] The target branch of this PR is as intended.
- [ ] The reviewer understands the issues in this PR.
- [ ] The reviewer plans to review with an appropriate workload based on the importance of this PR.

---
> When the PR reviewer concludes that this PR is ready to be merged, please fill in the checklist below by posting it in the comment. If there is more than one reviewer, please do this on your own.

#### Optional Entries
- [ ] The reviewer noted that if you believe that new tests are needed to evaluate this PR, they have been noted.
- [ ] If minor refactorings are not mentioned in the PR, I understand the intent.
- [ ] If this PR is a PR type A/B-1, we have agreed on this PR's design, direction, and granularity in the related issue.

#### Mandatory Entries
- [ ] The reviewer understands how this PR addresses the issue and the specific changes.
- [ ] This PR uses the best possible issue resolution that the reviewer can think of.
- [ ] This PR is ready to be merged.

retrage committed:
executor: Add `AFLExecutorInterface` class

This commit adds `AFLExecutorInterface` class to represent required
functions for an executor to support AFL-compatible algorithms. This
class accepts `std::shared_ptr<T>` of any class `T` as long as it
satisfies the function requirements. This behavior mimics Rust's trait
objects [1] using type erasure.

[1] https://doc.rust-lang.org/book/ch17-02-trait-objects.html

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

retrage committed:
build: Change to use `AFLExecutorInterface` for AFL family

This commit changes to use `AFLExecutorInterface` instead of using
`NativeLinuxExecutor` directly so that the algorithms are independent
from actual executor type. Note that this commit does not affect
algorithm behaviors and the executor is still fixed to
`NativeLinuxExecutor`.

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

retrage committed:
build: Change to use `LibFuzzerExecutorInterface` for libFuzzer family

This commti changes to use `LibFuzzerExecutorInterface` instead of using
`NativeLinuxExecutor` directly so that the algorithms are independent
form actual executor type. Note that this commit does not affect
algorithm behaviors and the executor is still fixed to
`NativeLinuxExecutor`.

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

Fadis commented on include/fuzzuf/executor/traits/afl_capable.hpp:
```diff
@@ -0,0 +1,105 @@
+  template<class T> DynAFLCapable(const T& executor)
+  : _container(new DynContainerDerived<T>(executor))
+  {}
+  ~DynAFLCapable() { delete _container; }
```
Since this class is copyable, this causes double free.

retrage replied:
Thanks, I added some deletion of implicit constructors.

potetisensei replied:
Using `std::unique_ptr` instead of raw pointer is safer and clearer in the first place? (is there any problem to do that?

retrage replied:
Fixed to use `std::unique_ptr` in [6532c86](https://github.com/fuzzuf/fuzzuf/pull/41/commits/6532c867cab29106390551e6f62aa905652fbbfd) and [16bb5ea](https://github.com/fuzzuf/fuzzuf/pull/41/commits/16bb5eaadfce9f423d019f2c9a6187b9ac375998).

potetisensei commented on include/fuzzuf/executor/traits/afl_capable.hpp:
```diff
@@ -0,0 +1,109 @@
+    }
+
+  private:
+    T _container;
```
It would be confusing that both `DynAFLCapable` and `DynContainerDerived` have the member `_container`. 
I suggest renaming `DynContainerDerived`'s `_container` to another name(like _executor or _value).

retrage replied:
Fixed with [6532c86](https://github.com/fuzzuf/fuzzuf/pull/41/commits/6532c867cab29106390551e6f62aa905652fbbfd) and [16bb5ea](https://github.com/fuzzuf/fuzzuf/pull/41/commits/16bb5eaadfce9f423d019f2c9a6187b9ac375998) by changing to `_executor`.

potetisensei commented on include/fuzzuf/executor/traits/afl_capable.hpp:
```diff
@@ -0,0 +1,109 @@
+    DynContainerDerived(T const &executor) : _container(executor) {}
+
+    void Run(const u8 *buf, u32 len, u32 timeout_ms=0) {
+      _container->Run(buf, len, timeout_ms);
```
Should the type of `DynContainerDerived<T>::_container` (= `T`) be pointer types?
If so, it must be noted in comments.
Also, if you can restrict the type (for example, allow only `std::shared_ptr<Foo>` as `T`), it would be better.
Accepting arbitrary pointer types as `T` means that we want to use at least two types among `Exector*`, `std::unique_ptr<Executor>`, and `std::shared_ptr<Executor>`.

retrage replied:
Fixed with [6532c86](https://github.com/fuzzuf/fuzzuf/pull/41/commits/6532c867cab29106390551e6f62aa905652fbbfd) and [16bb5ea](https://github.com/fuzzuf/fuzzuf/pull/41/commits/16bb5eaadfce9f423d019f2c9a6187b9ac375998) by restricting the type with `std::shared_ptr<T>`.

potetisensei commented on include/fuzzuf/executor/traits/afl_capable.hpp:
```diff
@@ -0,0 +1,109 @@
+ * @details The executor for AFL must have the methods declared in this class.
+ * This class is to perform type erasure for the executor class to abstract
+ * the executor on the algortihms. We found that the boost's type_erasure does
+ * not meet our needs, so we implemented our own type erasure.
```
Here, we should leave comments about the restriction imposed on `T`. 
(The dereferenced type of ) `T` should have 4 member functions right? Please describe what each function should do.
Of course you can instead write the explanation of each function directly above line 55, 59, 63, 67.

retrage replied:
Added comments for Doxygen in [60df237](https://github.com/fuzzuf/fuzzuf/pull/41/commits/60df23722f7f1ab268855245b1df770409eadd5f) and [c202523](https://github.com/fuzzuf/fuzzuf/pull/41/commits/c20252346ef38059b0dcdc0bd55a3f595d0fafdc).

potetisensei commented on include/fuzzuf/executor/traits/libfuzzer_capable.hpp:
```diff
@@ -0,0 +1,127 @@
+#include "fuzzuf/feedback/inplace_memory_feedback.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+/**
```
The same comments as `DynAFLCapable` We should deal with this class after the comments in `DynAFLCapable` are all resolved.

potetisensei commented on include/fuzzuf/executor/traits/afl_capable.hpp:
```diff
@@ -0,0 +1,109 @@
+ * the executor on the algortihms. We found that the boost's type_erasure does
+ * not meet our needs, so we implemented our own type erasure.
+ */
+class DynAFLCapable
```
I want to think of another name if possible because this name doesn't express that this is a container for "executors". But this is always difficult...

retrage replied:
I use the word `Capable` from Rust's trait naming convention (e.g. `Pausable` trait to support live migration on the VMM https://github.com/cloud-hypervisor/cloud-hypervisor/blob/main/vm-migration/src/lib.rs#L65). I'm unwilling to use `Executor` for these classes because they are requirements rather than actual executors.
What about using namespaces? For example, `executor::traits::AFLCapable` corresponds to `DynAFLCapable`. This clarifies the class is the executor's trait (container) that can be used for AFL-compatible algorithms.

potetisensei replied:
Indeed. Then I think we can leave this as it is.

potetisensei commented on include/fuzzuf/algorithms/nezha/create.hpp:
```diff
@@ -69,13 +69,14 @@ auto CreateRunSingleTarget(const fs::path &target_path,
       create_info.exec_timelimit_ms, create_info.exec_memlimit,
       create_info.forksrv, path_to_write_seed, create_info.afl_shm_size,
       create_info.bb_shm_size, true));
+  auto executor = std::make_unique<DynLibFuzzerCapable>(std::move(nle));
```
If I'm not mistaken, you need `DynLibFuzzerCapable::DynLibFuzzerCapable(T&&)` to make `std::move(nle)` actually move `nle`. Probably `T&&` would be casted to `const T&` because we don't define move constructors.

retrage replied:
I added move constructor in [5be878d](https://github.com/fuzzuf/fuzzuf/pull/41/commits/5be878d42a3206524a6732f7a10143dd4bb1abc8) and [bdacfde](https://github.com/fuzzuf/fuzzuf/pull/41/commits/bdacfde1fd6e315304ac1a44d86819b109ac7971). I'm not sure the classes accept copy constructors. Should we remove them?

potetisensei reviewed (changes requested):
Sorry for being late. I was off last Friday.
I have some points to discuss and left them as comments. So please check them.

Answering each of your concerns:

> 
> 1. Type Erasure Implementation
> 
> I implemented the classes with type erasure on my own. This is usually discouraged because bugs in my own implementation of type erasure are difficult to debug and can lead to "shooting myself in the foot". However, I found that the implementation using boost/type_erasure did not meet our needs in this case, as the amount of code using boost was not much different from my own implementation, and error messages tended to be more difficult. For this reason, I decided to implement type erasure myself. There is room for discussion on this policy.

Probably using boost/type_erasure is a better solution for those who are familiar with it. This is because its components called *concept* and *interface* would make clear what we want to define as a trait and where we implement it. However, considering there are many people who don't know about boost/type_erasure (even among us!), we would have to leave lots of comments about the things that is trivial to those who know it. Perhaps, ridiculously, the amount of comments can be even larger if we use boost/type_erasure and leave thoughtful comments, compared to our own implementation.  

Moreover, as you noted, depending on concepts we define, the amount of code would not be reduced, and compilation errors would be more difficult in boost/type_erasure. Therefore, I have no opposition to have our own type erasure implementation in this situation.



> 2. Performance
> 
> Because type erasure hides the actual type, runtime performance may suffer depending on compile-time optimizations. Although I have not yet measured and compared performance, I did not experience any performance degradation when running the unit test.
> 
> Please review the type erasure classes to ensure that no unnecessary copies are made. They take std::shared_ptr<TExecutor> as a constructor argument rather than executor itself. I believe that the executor is not created unnecessarily, but I am not sure if this implementation is correct.
> 

Putting it simply, what type erasure does is just realizing something like polymorphism(or dynamic dispatch) without inheritance. Hence, even if we used inheritance, defining a base class and derived classes to do the same as this PR, the performance of that implementation would be the same as the performance of this one. So I think we don't have to be worried too much (though of course we should take benchmarks if possible). 
Compared to template functions/classes, type erasure, as well as inheritance, will perform worse. But I think it must be more problematic that we have to write lots of `template<>` everywhere and will get a humongous binary as a result of compilation.

I suggested a fix for one unnecessary copy, but actually they don't affect the performance usually. This is because your concerning unnecessary copies are occurred in just `std::shared_ptr<T>`, not `T`itself. So they just unnecessarily take locks and increment refcounts at worst. Moreover, those copies would happen in initialization, not during a fuzzing campaign right?

> 3. Naming
> 
> These classes are placed under include/fuzzuf/executor/traits. I used the traits directory name because it is designed to be aware of Rust's trait objects, but it may cause misunderstanding because trait in C++ is a different concept from Rust's trait. If you have a better alternative directory name, please comment.

I think naming "trait" is somewhat relevant. As an alternative wording, "concept" can be used, but as you may have thought, concept is usually a more general idea describing the conditions for SFINAE or sometimes other things in BOOST and thus ambiguous. I think trait is nearer to what we mean by this design.

> The maintainer of this repository will set up a reviewer for each PR.
> PR reviewers should review this PR in terms of the checklist below before moving on to a detailed code review. Please comment on their initial response by filling in the checklist below.

#### Optional Entries
- [x] The reviewer assigned more reviewers if needed.
- [x] The reviewer noted that it is necessary to break out some of the changes in this PR into other PRs if needed.
- [x] The reviewer noted that the initial response is insufficient if needed.

#### Mandatory Entries
- [x] The title of this PR summarizes the changes made by this PR properly.
- [x] The target branch of this PR is as intended.
- [x] The reviewer understands the issues in this PR.
- [x] The reviewer plans to review with an appropriate workload based on the importance of this PR.

retrage committed:
cli: Add `fuzzuf::cli::fuzzer::nautilus` namespace

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>

potetisensei reviewed (approved):
LGTM

retrage committed:
executor: Add `LibFuzzerExecutorInterface` class

This commit adds `LibFuzzerExecutorInterface` class to represent
required functions for an executor to support libFuzzer-compatible
algorithms. This class accepts `std::shared_ptr<T>` of any class `T` as
long as it satisfies the function requirements. This behavior mimics
Rust's trait objects [1] using type erasure.

[1] https://doc.rust-lang.org/book/ch17-02-trait-objects.html

Signed-off-by: Akira Moroo <akiram@ricsec.co.jp>
